### PR TITLE
Advanced Quiz Mode

### LINKED ISSUE #28: As a user, I want to user  enter quiz mode, so that I can test my knowledge on a collection of flashcards
Some things to do:
- [x] Implement command to enter quiz mode. 
- [ ] Implement command to quit quiz mode.
- [ ] Implement command to give feedback on quiz mode.
- [x] Design the UI in the quiz mode. #26 
- [ ] Implement recommendation system based on statistics.

### LINKED ISSUE #45: Proper quiz mode
Currently, quiz mode selects random card and shows to the user both front and back face. User should be able to hide either the front face or the back face.

### LINKED ISSUE #47: As a user, I want to be able to quickly navigate through quiz mode
Currently, in quiz mode I need to type show, then type good/bad for each card, which is tedious. I want to be able to just press enter and it will automatically show the back face, or mark it as good if it has been shown.

### PR DESCRIPTION
**Pull request by sergiovieri**
This PR adds:
- Quiz mode that sorts cards by statistics
- Quiz panel that contains cards remaining, number of good and bad feedback
- User feedback through GoodCommand, BadCommand
- ShowCommand to show the back face of a card during quiz mode
- SmartCommand that executes ShowCommand if back face is not yet shown, or GoodCommand if back face is already shown

Fixes #45, fixes #47, and fixes #28

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/BadCommand.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import seedu.address.model.Model;
+
+/**
+ * Bad user feedback
+ */
+public class BadCommand extends Command {
+
+    public static final String COMMAND_WORD = "bad";
+
+    public static final String MESSAGE_BAD = "真可惜!";
```
Comment (lungsin) [inline src/main/java/seedu/address/logic/commands/BadCommand.java]: Change `MESSAGE_BAD` to english.

File: src/main/java/seedu/address/logic/commands/QuizCommand.java [hunk 2]
```diff
@@ -19,22 +19,47 @@
...
+
+        List<Flashcard> quizCards = getShuffledFlashCards(model);
+
+        if (quizCards.isEmpty()) {
+            throw new CommandException(MESSAGE_QUIZ_FAILURE_EMPTY);
+        }
+
+        model.resetQuizStat();
+        model.setQuizFlashcards(FXCollections.observableArrayList(quizCards));
+        model.setQuizMode(-1);
```
Comment (lungsin) [inline src/main/java/seedu/address/logic/commands/QuizCommand.java]: magic number -1.

File: src/main/java/seedu/address/logic/commands/QuizCommand.java [hunk 3]
```diff
@@ -19,22 +19,47 @@
...
     }
 
     @Override
     public CommandResult execute(Model model, CommandHistory history) throws CommandException {
         requireNonNull(model);
 
-        List<Flashcard> filteredPersonList = model.getFilteredFlashcardList();
-        Flashcard randomFlashcard = filteredPersonList.get((int) Math.floor(Math.random() * filteredPersonList.size()));
-        model.setSelectedFlashcard(randomFlashcard);
+        if (model.getQuizMode() != 0) {
```
Comment (lungsin) [inline src/main/java/seedu/address/logic/commands/QuizCommand.java]: `0` is magic number

File: src/main/java/seedu/address/logic/commands/ShowCommand.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package seedu.address.logic.commands;
+
+import static java.util.Objects.requireNonNull;
+
+import seedu.address.logic.CommandHistory;
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.Model;
+
+/**
+ * Good user feedback
```
Comment (lungsin) [inline src/main/java/seedu/address/logic/commands/ShowCommand.java]: I don't understand the javadoc.

File: src/main/java/seedu/address/logic/commands/ShowCommand.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    public static final String COMMAND_WORD = "show";
+
+    public static final String MESSAGE_SHOW = "Back face shown";
+    public static final String MESSAGE_FAILURE_NOT_QUIZ_MODE = "Cannot show card outside quiz mode";
+
+    @Override
+    public CommandResult execute(Model model, CommandHistory history) throws CommandException {
+        requireNonNull(model);
+
+        if (model.getQuizMode() == 0) {
```
Comment (lungsin) [inline src/main/java/seedu/address/logic/commands/ShowCommand.java]: magic number 0.

File: src/main/java/seedu/address/logic/commands/ShowCommand.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            throw new CommandException(MESSAGE_FAILURE_NOT_QUIZ_MODE);
+        }
+        model.setQuizMode(1);
```
Comment (lungsin) [inline src/main/java/seedu/address/logic/commands/ShowCommand.java]: This one as well. You can be a magician.

File: src/main/java/seedu/address/logic/commands/SmartCommand.java [hunk 7]
```diff
@@ -0,0 +1,30 @@
+package seedu.address.logic.commands;
+
+import static java.util.Objects.requireNonNull;
+
+import seedu.address.logic.CommandHistory;
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.Model;
+
+/**
+ * Good user feedback
```
Comment (lungsin) [inline src/main/java/seedu/address/logic/commands/SmartCommand.java]: javadoc :''

File: src/main/java/seedu/address/logic/commands/SmartCommand.java [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    @Override
+    public CommandResult execute(Model model, CommandHistory history) throws CommandException {
+        requireNonNull(model);
+
+        if (model.getQuizMode() == -1) {
+            return new ShowCommand().execute(model, history);
+        } else if (model.getQuizMode() == 1) {
+            return new GoodCommand().execute(model, history);
+        } else {
+            throw new CommandException(MESSAGE_FAILURE);
```
Comment (lungsin) [inline src/main/java/seedu/address/logic/commands/SmartCommand.java:29]: magician :'

File: src/test/java/systemtests/QuizCommandSystemTest.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    public void quiz() {
+        executeCommand("quiz");
+        assertEquals(-1, (int) getModelQuizMode());
+        executeCommand("show");
+        assertEquals(1, (int) getModelQuizMode());
+        int size = getModelQuizFlashcardSize();
+        executeCommand("good");
+        assertEquals(size - 1, getModelQuizFlashcardSize());
+        executeCommand("exit");
+        assertEquals(0, (int) getModelQuizMode());
```
Comment (lungsin) [inline src/test/java/systemtests/QuizCommandSystemTest.java]: magic number.

File: src/main/java/seedu/address/model/ModelManager.java [hunk 10]
```diff
@@ -213,6 +217,72 @@ private void ensureSelectedFlashcardIsValid(ListChangeListener.Change<? extends
...
+    }
+
+    @Override
+    public void setQuizMode(Integer quizMode) {
+        this.quizMode.setValue(quizMode);
+    }
+
+    @Override
+    public void showNextQuizCard() {
+        quizMode.setValue(-1);
```
Comment (lungsin) [inline src/main/java/seedu/address/model/ModelManager.java]: magic number

File: src/test/java/seedu/address/logic/commands/BadCommandTest.java [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+        new BadCommand().execute(model, commandHistory);
+        assertEquals((int) model.getQuizBad().getValue(), 1);
+    }
+
+    @Test
+    public void execute_emptyQuizFlashcards_exitsQuiz() throws CommandException {
+        new QuizCommand().execute(model, commandHistory);
+        model.getQuizFlashcards().clear();
+        new BadCommand().execute(model, commandHistory);
+        assertEquals(0, (int) model.getQuizMode());
```
Comment (lungsin) [inline src/test/java/seedu/address/logic/commands/BadCommandTest.java]: magic constant

File: src/test/java/seedu/address/logic/commands/ExitCommandTest.java [hunk 12]
```diff
@@ -19,4 +23,12 @@ public void execute_exit_success() {
         CommandResult expectedCommandResult = new CommandResult(MESSAGE_EXIT_ACKNOWLEDGEMENT, false, true);
         assertCommandSuccess(new ExitCommand(), model, commandHistory, expectedCommandResult, expectedModel);
     }
+
+    @Test
+    public void execute_inQuiz_exitsQuiz() throws CommandException {
+        Model newModel = new ModelManager(getTypicalCardCollection(), new UserPrefs());
+        new QuizCommand().execute(newModel, commandHistory);
+        new ExitCommand().execute(newModel, commandHistory);
+        assertEquals(0, (int) newModel.getQuizMode());
```
Comment (lungsin) [inline src/test/java/seedu/address/logic/commands/ExitCommandTest.java]: magic constant

File: src/test/java/seedu/address/logic/commands/GoodCommandTest.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+public class GoodCommandTest {
+
+    private Model model = new ModelManager(getTypicalCardCollection(), new UserPrefs());
+    private CommandHistory commandHistory = new CommandHistory();
+
+    @Test
+    public void execute_good_success() throws CommandException {
+        new QuizCommand().execute(model, commandHistory);
+        new GoodCommand().execute(model, commandHistory);
+        assertEquals((int) model.getQuizGood().getValue(), 1);
```
Comment (lungsin) [inline src/test/java/seedu/address/logic/commands/GoodCommandTest.java:35]: magic constant

File: src/test/java/seedu/address/logic/commands/GoodCommandTest.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+import seedu.address.model.UserPrefs;
+
+/**
+ * Contains integration tests and unit tests for GoodCommand.
+ */
+public class GoodCommandTest {
+
+    private Model model = new ModelManager(getTypicalCardCollection(), new UserPrefs());
+    private CommandHistory commandHistory = new CommandHistory();
+
```
Comment (lungsin) [inline src/test/java/seedu/address/logic/commands/GoodCommandTest.java:30]: add setUp method. 

```
     private Model model;
     private CommandHistory commandHistory;

      @Before
      public void setUp() {
            model = new ModelManager(getTypicalCardCollection(), new UserPrefs());
	    commandHistory = new CommandHistory();
      }
```

File: src/test/java/seedu/address/logic/commands/QuizCommandTest.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+import seedu.address.model.UserPrefs;
+
+/**
+ * Contains integration tests and unit tests for QuizCommand.
+ */
+public class QuizCommandTest {
+
+    private Model model = new ModelManager(getTypicalCardCollection(), new UserPrefs());
+    private CommandHistory commandHistory = new CommandHistory();
+
```
Comment (lungsin) [inline src/test/java/seedu/address/logic/commands/QuizCommandTest.java:30]: add setUp method.

```
     private Model model;
     private CommandHistory commandHistory;

      @Before
      public void setUp() {
            model = new ModelManager(getTypicalCardCollection(), new UserPrefs());
	    commandHistory = new CommandHistory();
      }
```

File: src/test/java/seedu/address/logic/commands/QuizCommandTest.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        command.execute(model, commandHistory);
+        assertEquals((int) model.getQuizMode(), -1);
+        assertEquals(model.getQuizFlashcards().size(), model.getFilteredFlashcardList().size() - 1);
+    }
+
+
+    @Test
+    public void execute_inQuizMode_failure() throws CommandException {
+        QuizCommand command = new QuizCommand();
+        model.setQuizMode(-1);
```
Comment (lungsin) [inline src/test/java/seedu/address/logic/commands/QuizCommandTest.java]: magician

File: src/test/java/seedu/address/logic/commands/ShowCommandTest.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+import seedu.address.model.ModelManager;
+import seedu.address.model.UserPrefs;
+
+/**
+ * Contains unit tests for ShowCommand.
+ */
+public class ShowCommandTest {
+    private Model model = new ModelManager(getTypicalCardCollection(), new UserPrefs());
+    private CommandHistory commandHistory = new CommandHistory();
+
```
Comment (lungsin) [inline src/test/java/seedu/address/logic/commands/ShowCommandTest.java:30]: add setUp method.

```
     private Model model;
     private CommandHistory commandHistory;

      @Before
      public void setUp() {
            model = new ModelManager(getTypicalCardCollection(), new UserPrefs());
	    commandHistory = new CommandHistory();
      }
```

File: src/test/java/seedu/address/logic/commands/ShowCommandTest.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+    @Test
+    public void execute_inQuiz_success() throws CommandException {
+        new QuizCommand().execute(model, commandHistory);
+        new ShowCommand().execute(model, commandHistory);
+        assertEquals((int) model.getQuizMode(), 1);
```
Comment (lungsin) [inline src/test/java/seedu/address/logic/commands/ShowCommandTest.java]: magic

File: src/test/java/seedu/address/logic/commands/SmartCommandTest.java [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+import seedu.address.model.UserPrefs;
+
+/**
+ * Contains integration tests and unit tests for SmartCommand.
+ */
+public class SmartCommandTest {
+
+    private Model model = new ModelManager(getTypicalCardCollection(), new UserPrefs());
+    private CommandHistory commandHistory = new CommandHistory();
+
```
Comment (lungsin) [inline src/test/java/seedu/address/logic/commands/SmartCommandTest.java:30]: add setUp method.

```
     private Model model;
     private CommandHistory commandHistory;

      @Before
      public void setUp() {
            model = new ModelManager(getTypicalCardCollection(), new UserPrefs());
	    commandHistory = new CommandHistory();
      }
```

File: src/test/java/seedu/address/logic/commands/SmartCommandTest.java [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+    @Test
+    public void execute_inQuiz_success() throws CommandException {
+        new QuizCommand().execute(model, commandHistory);
+        new SmartCommand().execute(model, commandHistory);
+        assertEquals((int) model.getQuizMode(), 1);
+        new SmartCommand().execute(model, commandHistory);
+        assertEquals((int) model.getQuizMode(), -1);
```
Comment (lungsin) [inline src/test/java/seedu/address/logic/commands/SmartCommandTest.java]: magic

**Review by lungsin — CHANGES_REQUESTED** [review]
Some issues: 
1. Some magic number. 
2. @Before annotation in Junit test

File: src/test/java/seedu/address/logic/commands/GoodCommandTest.java (hunk 13 above)
Comment (sergiovieri) [reply src/test/java/seedu/address/logic/commands/GoodCommandTest.java:35]: This is not magic constant >:(

**Review by lungsin — APPROVED** [review]

**Pull request merged**